File: gt7_server/core/raceEngine.js
/**
 * 🏎️ SRO GT7 RACE ENGINE — DOUBLE LOCK EDITION
 * 1. Включает PIT после 3 секунд стабильного нахождения в боксах.
 * 2. ДЕМПФЕР АНИМАЦИИ: Удерживает PIT, если ложный триггер трека (2) длится меньше 3 секунд.
 * 3. Удерживает ярко-зеленый статус OUT ровно 5 секунд после РЕАЛЬНОГО выезда.
 */

const pitMemory = {};

function processServerPitStates(pilotName, pitState) {
    const now = Date.now(); // Текущее время в миллисекундах

    // Инициализируем кэш пилота при первом заезде
    if (!pitMemory[pilotName]) {
        pitMemory[pilotName] = {
            currentState: "TRACK",    // "TRACK", "PIT_PENDING", "PIT", "TRACK_PENDING", "OUT"
            pitEntryTime: 0,          // Когда игра впервые прислала 1 или 3
            trackTriggerTime: 0,      // Когда игра прислала 2 во время анимации в питах
            exitTime: 0               // Когда пилот реально уехал на трассу (2)
        };
    }

    const mem = pitMemory[pilotName];

    // 🔴 ЗОНА ПИТ-ЛЕЙНА (Игра прислала статус 1 или 3)
    if (pitState === 1 || pitState === 3) {
        
        // Если пилот ехал по трассе или был в фазе OUT и пересёк черту пит-лейна
        if (mem.currentState === "TRACK" || mem.currentState === "OUT") {
            mem.currentState = "PIT_PENDING";
            mem.pitEntryTime = now; // Фиксируем точное время запроса на заезд
        }

        // Если пилот в режиме ожидания подтверждения заезда (3-секундный замок на въезде)
        if (mem.currentState === "PIT_PENDING") {
            const pendingDuration = (now - mem.pitEntryTime) / 1000;

            if (pendingDuration >= 3.0) {
                mem.currentState = "PIT";
            } else {
                return { ui_mode: "DELTA", badge_text: "" };
            }
        }

        // Если заезд подтвержден — намертво фиксируем красный бэдж PIT и сбрасываем ложные таймеры
        if (mem.currentState === "PIT" || mem.currentState === "TRACK_PENDING") {
            mem.currentState = "PIT";
            mem.trackTriggerTime = 0;
            return { ui_mode: "PIT", badge_text: "PIT" };
        }
    }

    // 🟢 ВЫЕЗД НА ТРАССУ (Игра прислала статус 2)
    if (pitState === 2) {
        
        // Если игра шлёт двойку, пока мы ждали 3 секунды на въезде — сбрасываем ложный дребезг флага
        if (mem.currentState === "PIT_PENDING") {
            mem.currentState = "TRACK";
            mem.pitEntryTime = 0;
            return { ui_mode: "DELTA", badge_text: "" };
        }

        // 🔥 ИСПРАВЛЕНИЕ ПРОСКОКА АНИМАЦИИ: 
        // Если игра прислала трек (2), но пилот до этого сидел в чистом статусе PIT — 
        // мы НЕ включаем OUT сразу! Мы включаем режим проверки ложного выезда TRACK_PENDING!
        if (mem.currentState === "PIT") {
            mem.currentState = "TRACK_PENDING";
            mem.trackTriggerTime = now; // Начинаем отсчет 3-х секунд удержания PIT
        }

        // Пока мы находимся внутри 3-секундного демпфера анимации
        if (mem.currentState === "TRACK_PENDING") {
            const trackDuration = (now - mem.trackTriggerTime) / 1000;

            if (trackDuration < 3.0) {
                // 🛑 МЕНЬШЕ 3 СЕКУНД: Это тот самый баг ракурсов игры на анимации подъезда!
                // Жёстко удерживаем плашку PIT и не даем ей переключиться на OUT или TRACK!
                return { ui_mode: "PIT", badge_text: "PIT" };
            } else {
                // 🔥 БОЛЬШЕ 3 СЕКУНД: Сигнал трека стабилен, пилот РЕАЛЬНО покинул боксы!
                mem.currentState = "OUT";
                mem.exitTime = now; // Фиксируем точную миллисекунду реального выезда
            }
        }

        // Режим удержания ярко-зеленого статуса OUT ровно 5 секунд после реального выезда
        if (mem.currentState === "OUT") {
            const timePassed = (now - mem.exitTime) / 1000;

            if (timePassed < 5.0) {
                return { ui_mode: "OUT", badge_text: "OUT" };
            } else {
                // 5 секунд вышло — плавно гасим плашку, возвращаем чистый режим
                mem.currentState = "TRACK";
                mem.exitTime = 0;
                mem.pitEntryTime = 0;
                mem.trackTriggerTime = 0;
                return { ui_mode: "DELTA", badge_text: "" };
            }
        }
    }

    return { ui_mode: "DELTA", badge_text: "" };
}

module.exports = { processServerPitStates };

File: gt7_server/core/race_engine.py
import time

# Чистая оперативная память состояний для каждого пилота лиги
PIT_MEMORY = {}

def process_server_pit_clocks(pilot_name: str, pit_state: int, current_timestamp: float, current_badge_text: str) -> tuple:
    """
    ЭТАЛОННЫЙ СУДЕЙСКИЙ ДВИЖОК С ПАУЗОЙ ТАЙМЕРА (БЕЗ СИНТАКСИЧЕСКИХ ОШИБОК):
    - Выход из пита только после 5 секунд непрерывного статуса трассы (2) или выезда (3).
    - После подтверждения выезда — плашка TRACK горит 10 секунд, а таймер пита встает на ПАУЗУ.
    - Полная защита от дисконнектов (AFK) и простоев больше 120 секунд.
    """
    now = time.time()

    # 1. 🚨 СУДЕЙСКИЙ AFK ХУК: Жесткий дисконнект игры
    if pit_state == -1:
        return "AFK", "AFK", "AFK"

    # Инициализируем кэш пилота при первом появлении в сессии
    if pilot_name not in PIT_MEMORY:
        PIT_MEMORY[pilot_name] = {
            "current_state": "TRACK",     # Текущие состояния оверлея: TRACK, IN_BOX, ON_TRACK
            "entry_time": 0.0,            # Время въезда в боксы
            "exit_time": 0.0,             # Время реального выезда на трассу
            "track_status_start": 0.0,    # Таймер проверки 5 секунд стабильности трассы
            "paused_pit_duration": 0      # Замороженное время пита для вывода на плашку TRACK
        }

    mem = PIT_MEMORY[pilot_name]

    # =========================================================================
    # 🧮 ЛОГИКА СУДЕЙСКОГО ЗАМКА, ТАЙМ-АУТОВ И ЗАМОРОЗКИ ТАЙМЕРА НА ПАУЗУ
    # =========================================================================

    # А) ЕСЛИ ПРИСТАВКА ОФИЦИАЛЬНО ШЛЕТ ФЛАГ БОКСОВ (pit_state == 1)
    if pit_state == 1:
        if mem["current_state"] != "IN_BOX":
            mem["current_state"] = "IN_BOX"
            mem["entry_time"] = now
        
        mem["track_status_start"] = 0.0
        
        pit_duration = int(now - mem["entry_time"])
        if pit_duration > 120: 
            return "AFK", "AFK", "AFK"
            
        return "IN_BOX", "PIT", f"PIT {pit_duration}s"

    # Б) ЖЕСТКИЙ СИНТАКСИЧЕСКИЙ ФИКС: Проверяем стабильность, если статус равен 2 или 3
    elif pit_state in (2, 3):
        
        # Если мы до этого ехали по трассе в штатном режиме — просто катим дальше
        if mem["current_state"] == "TRACK":
            return "DELTA", "TRACK", ""

        # Если мы были в боксах (IN_BOX) и прилетела двойка/тройка от приставки
        if mem["current_state"] == "IN_BOX":
            if mem["track_status_start"] == 0.0:
                mem["track_status_start"] = now
            
            track_stable_duration = now - mem["track_status_start"]
            
            if track_stable_duration < 5.0:
                # 🛑 ЗАМОК АНИМАЦИИ: 5 секунд стабильной трассы еще не прошло! Таймер крутим дальше!
                pit_duration = int(now - mem["entry_time"])
                return "IN_BOX", "PIT", f"PIT {pit_duration}s"
            else:
                # 🟢 ВЫЕЗД ПОДТВЕРЖДЕН: Статус трассы удержался дольше 5 секунд подряд!
                mem["current_state"] = "ON_TRACK"
                mem["exit_time"] = now
                mem["paused_pit_duration"] = int(now - mem["entry_time"] - 5.0)

        # В) РЕЖИМ ВЫЕЗДА "ON_TRACK": ДЕРЖИМ ПЛАШКУ И ТАЙМЕР НА ПАУЗЕ 10 СЕКУНД
        if mem["current_state"] == "ON_TRACK":
            if now - mem["exit_time"] < 10.0:
                return "ON_TRACK", "TRACK", f"PIT {mem['paused_pit_duration']}s"
            else:
                mem["current_state"] = "TRACK"
                mem["track_status_start"] = 0.0
                return "DELTA", "TRACK", ""

    return "DELTA", "TRACK", ""
